### PR TITLE
Gif me node 22

### PR DESCRIPTION
**Pull request by danblackadder**

### CODE REVIEW

**Comment by edenchazard** [comment]
IT'S HAPPENING LADS

:haato_shake_if_it_existed_on_github:

File: server/workers/shareScrollWorker.ts [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+    const SAMPLE_FLAIR = 'saxifrage';
+    const SAMPLE_WEEKLY_CLICKS = 12345;
+    const SAMPLE_ALLTIME_CLICKS = 123456;
+    const { bannerBuffer, bannerWidth, bannerHeight } = await getBannerBase(
+      username,
+      dragonCount,
+      growingCount,
+      SAMPLE_WEEKLY_CLICKS,
+      SAMPLE_ALLTIME_CLICKS,
+      SAMPLE_FLAIR
```
Comment (edenchazard) [inline server/workers/shareScrollWorker.ts:136]: enough to make a man cry 😭

File: server/workers/shareScrollWorker.ts [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+
+async function generateBannerToTemporary(
+  username: string,
+  filePath: string,
+  token: string
+) {
+  // missing things currently:
+  // - complete the banner generation chain, one func at a time
+  // - get access token from user so we can count a hidden scroll's dragons
+  // - how to hit the hatchery's api for click info? flair info and images?
```
Comment (edenchazard) [inline server/workers/shareScrollWorker.ts:113]: I think the easiest approach will be passing it down to the worker instead of trying to import the db in the worker thread. that also then gives you the opportunity to validate the user is actually registered with the garden.

File: server/routes/share/scroll/[username].ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+  setHeaders(event, {
+    'Content-Type': 'image/webp',
+  });
+
+  return sendStream(event, createReadStream(
+    '/src/public/banner/in_progress.webp'
+  ));
+  // little thing: the url ends in .gif but this resource is a .webp.
+  // it serves just fine. will the filetype discrepancy be a problem later?
```
Comment (edenchazard) [inline server/routes/share/scroll/[username].ts:72]: I dunno. would need to check the `mime_type` though even that can't be trusted

File: server/routes/share/scroll/[username].ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    setHeaders(event, {
+      'Content-Type': 'image/gif',
+      'Cache-Control': `public, max-age=${expiry / 1000}`,
+    });
+
+    return sendStream(event, createReadStream(filePath));
+  }
+
+  // todo: what if banner generation fails (nonexistent scrollname)?
+  // i want to render the 'scroll not found' banner in that case
```
Comment (edenchazard) [inline server/routes/share/scroll/[username].ts:61]: So here you're probably going to want to take the validated scroll name and then check it against the user table. you can also fetch their flair, click amounts, etc at this point and pass it to the job (sort of mentioned this further down) at the same time.

File: server/workers/shareScrollWorker.ts [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+      .png()
+      .toBuffer();
+
+    console.log(
+      `Growing images generated in ${(performance.now() - startTime).toFixed(0)}ms`
+    );
+    return {
+      stripBuffer,
+      stripWidth: totalWidth,
+      stripHeight: HEIGHT,
```
Comment (edenchazard) [inline server/workers/shareScrollWorker.ts:365]: let's try and keep to thisSortOfCase pls

File: server/workers/shareScrollWorker.ts [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+    };
+  } catch (error) {
+    console.error('Error in getDragonStrip:', error);
+    throw error;
+  }
```
Comment (edenchazard) [inline server/workers/shareScrollWorker.ts:370]: I'm not sure this serves any purpose - catching an error and then rethrowing it.

File: server/workers/shareScrollWorker.ts [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+      const metadata = dragonMetadatas[i];
+
+      composites.push({
+        input: await dragonImage.toBuffer(),
+        top: HEIGHT - (metadata.height ?? 0),
+        left: xOffset,
+      });
+
+      xOffset += (metadata.width ?? 0) + 1;
+    }
```
Comment (edenchazard) [inline server/workers/shareScrollWorker.ts:352]: candidate for `Promise.all`

File: server/workers/shareScrollWorker.ts [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+          );
+        }
+        const arrayBuffer = await response.arrayBuffer();
+        return Buffer.from(arrayBuffer);
+      })
+    );
+
+    const dragonImages = dragonBuffers.map((buffer) => sharp(buffer));
+
+    const dragonMetadatas = await Promise.all(
```
Comment (edenchazard) [inline server/workers/shareScrollWorker.ts:323]: nice!

File: server/workers/shareScrollWorker.ts [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+      dragonIds.map(async (dragonId) => {
+        const response = await fetch(
+          `https://dragcave.net/image/${dragonId}.gif`
+        );
+        if (!response.ok) {
+          throw new Error(
+            `Failed to fetch dragon image ${dragonId}: ${response.statusText}`
+          );
+        }
+        const arrayBuffer = await response.arrayBuffer();
```
Comment (edenchazard) [inline server/workers/shareScrollWorker.ts:316]: I want to _say_ fetch already throws its own error I think. unless you need to see this specifically...?

File: server/workers/shareScrollWorker.ts [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+    const { height: usernameHeight, width: usernameWidth } =
+      await sharp(usernamePng).metadata();
+    composites.push({
+      input: usernamePng,
+      top: 9 + (16 - usernameHeight),
+      left: 118,
+    });
+
+    // flair
+    if (flair && flair !== '') {
```
Comment (edenchazard) [inline server/workers/shareScrollWorker.ts:202]:
```suggestion
    if (flair !== '') {
```

the first condition isn't needed

File: server/workers/shareScrollWorker.ts [hunk 11]
```diff
@@ -0,0 +1,548 @@
...
+      composites.push({
+        input: await flairImage.toBuffer(),
+        left: 118 + usernameWidth + 4,
+        top: 16 - Math.floor((flairHeight ?? 0) / 2),
+      });
+    }
+
+    const statText = (statName: string, statNumber: number) => `
+      <tspan fill="#dff6f5">${statName}:</tspan> 
+      <tspan fill="#f2bd59">${statNumber.toLocaleString()}</tspan>
```
[the remark this answers is not in the document]
Comment (endulum) [reply server/workers/shareScrollWorker.ts:231]: Okay so I notice I did this wrong - it works on numbers, not strings, so I should have `parseInt`'d that first - but `toLocaleString` on numbers [allows for commas](https://stackoverflow.com/questions/2901102/how-to-format-a-number-with-commas-as-thousands-separators/17663871#17663871), i.e. `12,345` instead of `12345`.

If you don't care for commas though, that's alright.

File: server/workers/shareScrollWorker.ts (hunk 7 above)
Comment (endulum) [reply server/workers/shareScrollWorker.ts:352]: Doesn't `Promise.all()` run everything in parallel? And this feels like something that should remain sequential, especially regarding the interaction with `xOffset`. Hnnn, maybe...

File: server/workers/shareScrollWorker.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (edenchazard) [reply server/workers/shareScrollWorker.ts:231]: @endulum this is a fair point. `toLocaleString` uses the system's locale by default which can change, though unlikely in our case. pass in `en-gb` or something so it's certain.

File: server/workers/shareScrollWorker.ts (hunk 6 above)
Comment (endulum) [reply server/workers/shareScrollWorker.ts:382]: Blocks [like these](https://github.com/edenchazard/garden-of-eden/blob/3df5fe0331dfc493a3eec4d6083e1c127a31ffb9/server/workers/shareScrollWorker.ts#L224C3-L227C4) were present before I worked on it. In its defense I do feel telling us where in the generation process an error occurred can be convenient in case the stack trace alone isn't helpful.

File: server/workers/shareScrollWorker.ts (hunk 9 above)
Comment (endulum) [reply server/workers/shareScrollWorker.ts:328]: The fetch can still return a not-okay response without erroring, like a 404. Unless I'm missing something obvious?

File: server/workers/shareScrollWorker.ts (hunk 7 above)
Comment (edenchazard) [reply server/workers/shareScrollWorker.ts:363]: it returns a promise that's resolved once the passed promises all succeed.  I dunno, feels like there's way here to precompute the `x offset` and then await all the buffers.

Comment (endulum) [reply server/workers/shareScrollWorker.ts]: How's this?
```js
const xOffsets: number[] = [0];
dragonMetadatas.forEach(metadata => {
  const prevOffset = xOffsets[xOffsets.length - 1];
  xOffsets.push(prevOffset + (metadata.width ?? 0) + 1);
});
    
await Promise.all(dragonImages.map(async (dragonImage, index) => {
  composites.push({
    input: await dragonImage.toBuffer(),
    top: stripHeight - (dragonMetadatas[index].height ?? 0),
    left: xOffsets[index]
  });
}));
```

Comment (edenchazard) [reply server/workers/shareScrollWorker.ts]: Nice, it's only a minor op anyway

File: server/workers/shareScrollWorker.ts (hunk 9 above)
Comment (edenchazard) [reply server/workers/shareScrollWorker.ts]: reasonable

File: server/routes/share/scroll/[username].ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+  // second, get user clicks
+  const now = DateTime.now();
+  const weekStart = now.startOf('week');
+  const { length: weeklyClicks } = await db
+    .select()
+    .from(clicksTable)
+    .where(
+      and(
+        gte(clicksTable.clicked_on, weekStart.toJSDate()),
+        eq(clicksTable.user_id, userId[0].id)
```
Comment (edenchazard) [inline server/routes/share/scroll/[username].ts:88]: you're missing a performance trick here with something called a `composite index`: BTREE indexes work left to right, meaning **you have to test the left most sides** to make use of the next condition.

we have a single index on `clicked_on`, but you have two conditions here. your query will use this index.

we can do better, though. there's a composite on `user_id`, `clicked_on`, so swap these two conditions and now your query will make use of that index and scan far less rows.

File: server/routes/share/scroll/[username].ts [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+      )
+    );
+  const { length: allTimeClicks } = await db
```
Comment (edenchazard) [inline server/routes/share/scroll/[username].ts:91]: 100% promise.all these

File: server/routes/share/scroll/[username].ts [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    .select()
+    .from(clicksTable)
+    .where(
+      eq(clicksTable.user_id, userId[0].id)
+    )
+  // these are almost the same query, how do i keep this a little drier?
```
Comment (edenchazard) [inline server/routes/share/scroll/[username].ts:97]: don't worry about it

File: server/routes/share/scroll/[username].ts [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+  // get flair
+  const [{ url: flairUrl }] = await db
+    .select({
+      url: itemsTable.url
+    })
+    .from(userTable)
+      .leftJoin(userSettingsTable, eq(userTable.id, userSettingsTable.user_id))
```
Comment (edenchazard) [inline server/routes/share/scroll/[username].ts:105]:
```suggestion
      .innerJoin(userSettingsTable, eq(userTable.id, userSettingsTable.user_id))
```

we 100% guaranteed that what exists in the left column will exist in the right. there's no possibility a user can exist without being present in the settings table. It'll also convince drizzle.

File: server/routes/share/scroll/[username].ts [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+      .leftJoin(itemsTable, eq(userSettingsTable.flair_id, itemsTable.id))
```
Comment (edenchazard) [inline server/routes/share/scroll/[username].ts:106]: yes, nice. perfect use of a left join here

File: server/routes/share/scroll/[username].ts [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+  // first, does user exist?
+  const userId = await db
+  // one could shorten this to `[{ id }]` for convenience like i saw you do elsewhere
+  // but this could throw an error if a user can't actually be found.
+  // or maybe we can get around that by wrapping all this db-touching in a try/catch?
+    .select({
+      id: userTable.id,
+    })
+    .from(userTable)
+    .where(eq(userTable.username, params.username));
```
Comment (edenchazard) [inline server/routes/share/scroll/[username].ts:72]: reminder for myself: add a unique index to username now

File: server/routes/share/scroll/[username].ts [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+        return dotIndex !== -1 ? v.substring(0, dotIndex) : v;
+      }),
+  });
+  const params = await getValidatedRouterParams(event, schema.parse);
+
+  // first, does user exist?
+  const userId = await db
+  // one could shorten this to `[{ id }]` for convenience like i saw you do elsewhere
+  // but this could throw an error if a user can't actually be found.
+  // or maybe we can get around that by wrapping all this db-touching in a try/catch?
```
[the remark this answers is not in the document]
Comment (endulum) [reply server/routes/share/scroll/[username].ts:67]: Declaring with `const [{ id }] = await db...` and running a nonexistent scrollname against the route yields a Nuxt error:
```
hatchery-1  |  ERROR  [nuxt] [request error] [unhandled] [500] Cannot read properties of undefined (reading 'id')
hatchery-1  |   at Object.handler (./server/routes/share/scroll/[username].ts:63:1)
...
```

[the remark this answers is not in the document]
Comment (edenchazard) [reply server/routes/share/scroll/[username].ts:67]: interesting

[the remark this answers is not in the document]
Comment (endulum) [reply server/routes/share/scroll/[username].ts:67]: Oh, but `const [user] = ...` works though, and can safely return undefined.

File: server/routes/share/scroll/[username].ts (hunk 13 above)
Comment (edenchazard) [reply server/routes/share/scroll/[username].ts]: take a look at `statistics/index.ts` you can inline the individual promises as one big boi chunk or break it into async functions.

File: server/routes/share/scroll/[username].ts [hunk 19]
```diff
@@ -24,7 +25,26 @@ const getWeeklyClicks = async (userId: number) => {
       eq(clicksTable.user_id, userId),
       gte(clicksTable.clicked_on, weekStart.toJSDate()),
     ));
-  return weeklyClicks
+
+  const [ weeklyRank ]: { rank: number }[] = await db
```
Comment (edenchazard) [inline server/routes/share/scroll/[username].ts:29]: you shouldn't need to type hint this

File: server/routes/share/scroll/[username].ts [hunk 20]
```diff
@@ -24,7 +25,26 @@ const getWeeklyClicks = async (userId: number) => {
...
+    .select({
+      rank: clicksLeaderboardTable.rank
+    })
+    .from(clicksLeaderboardTable)
+      .innerJoin(userTable, eq(userTable.id, clicksLeaderboardTable.user_id))
+    .where(
+      and(
+        eq(userTable.id, userId),
+        eq(clicksLeaderboardTable.leaderboard, 'weekly')
```
Comment (edenchazard) [inline server/routes/share/scroll/[username].ts:38]: swap these for indexing reasons

File: workers/shareScrollWorker.ts [hunk 21]
```diff
@@ -0,0 +1,537 @@
...
+    return await compositeImage.composite(composites).png().toBuffer();
+  } catch (error) {
+    console.error('Error in getBannerBase:', error);
+    throw error;
+  }
+}
+
+async function getDragonStrip(dragonIds: string[], clientSecret: string) {
+  try {
+    let validDragonIds: string[] = [];
```
Comment (edenchazard) [inline workers/shareScrollWorker.ts:337]:
```suggestion
    const validDragonIds: string[] = [];
```

File: workers/shareScrollWorker.ts [hunk 22]
```diff
@@ -0,0 +1,537 @@
...
+        timeout: 10000,
+      }
+    );
+
+    if (errors.length > 0) {
+      throw new Error('Errors getting bulk dragons: ' + errors);
+    } else {
+      validDragonIds = Object.keys(dragons).filter((key) => {
+        return 'hoursleft' in dragons[key] && dragons[key].hoursleft > 0;
+      });
```
Comment (edenchazard) [inline workers/shareScrollWorker.ts:360]:
```suggestion
      validDragonIds.push(...Object.keys(dragons).filter((key) => 'hoursleft' in dragons[key] && dragons[key].hoursleft > 0));
```

File: workers/shareScrollWorker.ts [hunk 23]
```diff
@@ -0,0 +1,537 @@
...
+        const response = await ofetch(
+          `https://dragcave.net/image/${dragonId}.gif`,
+          {
+            retry: 3,
+            retryDelay: 1000,
+            timeout: 10000,
+          }
+        );
+        return Buffer.from(await response.arrayBuffer());
+        // i can't find ofetch documentation for the proper way to fetch as buffer
```
Comment (edenchazard) [inline workers/shareScrollWorker.ts:381]: https://github.com/unjs/ofetch?tab=readme-ov-file#%EF%B8%8F-parsing-response

File: workers/shareScrollWorker.ts [hunk 24]
```diff
@@ -0,0 +1,537 @@
...
+        retryDelay: 1000,
+        timeout: 10000,
+      }
+    );
+
+    if (errors.length > 0) {
+      throw new Error('Errors getting bulk dragons: ' + errors);
+    } else {
+      validDragonIds = Object.keys(dragons).filter((key) => {
+        return 'hoursleft' in dragons[key] && dragons[key].hoursleft > 0;
```
Comment (edenchazard) [inline workers/shareScrollWorker.ts:359]:
```suggestion
        return dragons[key]?.hoursleft > 0;
```

File: server/routes/share/scroll/[request].ts [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+      'Cache-Control': `public, max-age=${useRuntimeConfig().bannerCacheExpiry}`,
+    });
+    return sendStream(event, createReadStream(filePath));
+  }
+
+  setHeader(event, 'Content-Type', contentType);
+
+  return sendStream(
+    event,
+    createReadStream(`/src/resources/banner/in_progress${params.data.ext}`)
```
[the remark this answers is not in the document]
Comment (endulum) [reply server/routes/share/scroll/[request].ts:176]: LMAO will try.

File: server/routes/share/scroll/[request].ts [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+
+  const filePath = `/cache/scroll/${params.data.id}${params.data.ext}`;
+  const contentType = contentTypes[params.data.ext];
+  const user = await getUser(params.data.id);
+
+  if (!user) return sendNotFound(event);
+
+  // todo later: it might be better to contain all of these variables
+  // into a "data" object instead of threading every single one
+  // through multiple funcs.
```
[the remark this answers is not in the document]
Comment (endulum) [reply server/routes/share/scroll/[request].ts:161]: Wait... what's the suggestion here exactly?

[the remark this answers is not in the document]
Comment (edenchazard) [reply server/routes/share/scroll/[request].ts:161]: removing the comments because they no longer serve purpose at this scope, unless you move it to `sendJob`.

[the remark this answers is not in the document]
Comment (endulum) [reply server/routes/share/scroll/[request].ts:161]: Oh, I glossed over the red.

File: database/schema.ts [hunk 27]
```diff
@@ -292,6 +292,38 @@ export const purchasesTable = mysqlTable('purchases', {
...
+  flair_name: varchar('flair_name', {
+    length: 24,
+  }),
+  stat_gen_time: mediumint('stat_gen_time', { unsigned: true }),
+  dragon_fetch_time: mediumint('dragon_fetch_time', { unsigned: true }),
+  dragon_gen_time: mediumint('dragon_gen_time', { unsigned: true }),
+  frame_gen_time: mediumint('frame_gen_time', { unsigned: true }),
+  gif_gen_time: mediumint('gif_gen_time', { unsigned: true }),
+  total_time: mediumint('total_time', { unsigned: true }),
+  // error: what format should error be, since it can be anything? just a string?
```
Comment (edenchazard) [inline database/schema.ts:321]: I'm guessing `json`

File: database/schema.ts [hunk 28]
```diff
@@ -292,6 +292,38 @@ export const purchasesTable = mysqlTable('purchases', {
...
+    unsigned: true,
+  })
+    .references(() => userTable.id, {
+      onDelete: 'cascade',
+    })
+    .notNull(),
+  username: varchar('username', {
+    length: 32,
+  }).notNull(),
+  flair_name: varchar('flair_name', {
```
Comment (edenchazard) [inline database/schema.ts:312]:
```suggestion
  flair_path: varchar('flair_path', {
```

File: database/schema.ts [hunk 29]
```diff
@@ -292,6 +292,38 @@ export const purchasesTable = mysqlTable('purchases', {
...
 });
 
+export const bannerJobTable = mysqlTable('banner_jobs', {
+  id: bigint('id', {
+    unsigned: true,
+    mode: 'number',
+  })
+    .autoincrement()
+    .primaryKey(),
+  user_id: mediumint('user_id', {
```
Comment (edenchazard) [inline database/schema.ts:302]:
```suggestion
  userId: mediumint('user_id', {
```

would really prefer it like this but I haven't really kept to this either especially early on so dw about it too much

File: database/schema.ts [hunk 30]
```diff
@@ -292,6 +292,38 @@ export const purchasesTable = mysqlTable('purchases', {
     .notNull(),
 });
 
+export const bannerJobTable = mysqlTable('banner_jobs', {
+  id: bigint('id', {
+    unsigned: true,
+    mode: 'number',
+  })
+    .autoincrement()
+    .primaryKey(),
```
Comment (edenchazard) [inline database/schema.ts:301]: think there's a nice alias for all this called `serial`

File: database/schema.ts [hunk 31]
```diff
@@ -292,6 +292,38 @@ export const purchasesTable = mysqlTable('purchases', {
     .notNull(),
 });
 
+export const bannerJobTable = mysqlTable('banner_jobs', {
```
Comment (edenchazard) [inline database/schema.ts:295]:
```suggestion
export const bannerJobsTable = mysqlTable('banner_jobs', {
```

File: database/schema.ts
Comment (edenchazard) [inline database/schema.ts:1]: I'd also add json columns for: `omitted` (with reason, perhaps) and `dragons`

File: database/schema.ts [hunk 32]
```diff
@@ -292,6 +292,38 @@ export const purchasesTable = mysqlTable('purchases', {
...
+  user_id: mediumint('user_id', {
+    unsigned: true,
+  })
+    .references(() => userTable.id, {
+      onDelete: 'cascade',
+    })
+    .notNull(),
+  username: varchar('username', {
+    length: 32,
+  }).notNull(),
```
Comment (edenchazard) [inline database/schema.ts:311]: I was going to flag this because we have the user id, but then I realised it's legitimate they can change their username and we want to log the data "as is". so I think you're right here, nice job

File: database/schema.ts [hunk 33]
```diff
@@ -292,6 +292,38 @@ export const purchasesTable = mysqlTable('purchases', {
...
+    length: 24,
+  }),
+  stat_gen_time: mediumint('stat_gen_time', { unsigned: true }),
+  dragon_fetch_time: mediumint('dragon_fetch_time', { unsigned: true }),
+  dragon_gen_time: mediumint('dragon_gen_time', { unsigned: true }),
+  frame_gen_time: mediumint('frame_gen_time', { unsigned: true }),
+  gif_gen_time: mediumint('gif_gen_time', { unsigned: true }),
+  total_time: mediumint('total_time', { unsigned: true }),
+  // error: what format should error be, since it can be anything? just a string?
+  generated_on: datetime('generated_on', { mode: 'date' })
```
Comment (edenchazard) [inline database/schema.ts:322]:
```suggestion
  createdAt: datetime('created_at', { mode: 'date' })
```

just a more standardised descriptor

File: drizzle/0025_ambiguous_lady_ursula.sql
Comment (edenchazard) [inline drizzle/0025_ambiguous_lady_ursula.sql:1]: Once you're happy with how your schema is, you can drop all three of these migration files via `npx drizzle-kit drop` then do a new migration just to consolidate it all.

File: workers/shareScrollWorker.ts [hunk 34]
```diff
@@ -0,0 +1,593 @@
...
+    filePath,
+    weeklyClicks,
+    weeklyRank,
+    allTimeClicks,
+    allTimeRank,
+    dragons,
+    clientSecret,
+  } = message;
+
+  let performanceData: PerformanceData = {
```
Comment (edenchazard) [inline workers/shareScrollWorker.ts:44]:
```suggestion
  const performanceData: PerformanceData = {
```

File: workers/shareScrollWorker.ts [hunk 35]
```diff
@@ -0,0 +1,593 @@
...
+    // initialize every timing stat with null, filling them in
+    // one by one after each of the funcs do their things.
+    // let's use null for funcs that never ran (in case of error),
+    // and 0 for funcs that needn't run (in case of no carousel/frames)?
+    // this way, if we look at the row of a job that errored,
+    // if the error text alone doesn't help us locate it,
+    // we can still deduce from the null columns which function errored
```
Comment (edenchazard) [inline workers/shareScrollWorker.ts:51]: clever 👍

File: workers/shareScrollWorker.ts [hunk 36]
```diff
@@ -0,0 +1,593 @@
...
+      performanceData,
+      user,
+      filePath,
+      weeklyClicks,
+      weeklyRank,
+      allTimeClicks,
+      allTimeRank,
+      dragons,
+      clientSecret
+    );
```
Comment (edenchazard) [inline workers/shareScrollWorker.ts:75]: be cool to do `Object.assign` here so you can utilise `const`

File: workers/shareScrollWorker.ts [hunk 37]
```diff
@@ -0,0 +1,593 @@
...
+            },
+            tile: true,
+            blend: 'dest-in',
+          },
+          { input: flairPath, left: -1, top: -1 },
+          // actually, this cuts off the top and left
+          // of the flair image. fiddling with the `left` and `top`
+          // of either composite input did not keep the shadow.
+          // if only there was native drop shadow support.
+          // i guess we might drop the flair shadow? it's very subtle anyway
```
Comment (edenchazard) [inline workers/shareScrollWorker.ts:250]: may be worth dropping it then

File: database/schema.ts (hunk 27 above)
Comment (endulum) [reply database/schema.ts:321]: I feel a little safer just stringifying it since errors can be anything and won't always necessarily be valid JSON.

File: database/schema.ts
Comment (endulum) [reply database/schema.ts:1]: Can do but possibly without the reason? The check for growing is only one condition (`hoursLeft` is a pos integer)

File: database/schema.ts (hunk 30 above)
Comment (endulum) [reply database/schema.ts:301]: Encountering a syntax issue and it's likely related to [this fresh bug](https://github.com/drizzle-team/drizzle-orm/issues/3333). Gonna circumvent this for now.

Comment (endulum) [reply database/schema.ts:301]: Ah, wait, even going back on this change, the migrate is still bugging. Hm.

Comment (endulum) [reply database/schema.ts:302]: Okay, handling and migrating your schema suggestions one at a time it appears `serial` alone is in fact the problem.

Setting `id` to `serial('id')` as per the [doc example](https://orm.drizzle.team/docs/column-types/mysql#serial) yields an error on migrate: `Incorrect table definition; there can be only one auto column and it must be defined as a key`

Adding `.primaryKey()` yields a different parse error: `You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'AUTO_INCREMENT NOT NULL' at line 1` with the highlighted sql being `ALTER TABLE `banner_jobs` MODIFY COLUMN `id` serial AUTO_INCREMENT NOT NULL;`

If it seems like I'm blindly cobbling things together it's likely because I am ^^'

File: drizzle/0025_ambiguous_lady_ursula.sql
Comment (endulum) [reply drizzle/0025_ambiguous_lady_ursula.sql:1]: I do understand what's going on in migration - in migrating, drizzle runs these sql files sequentially to build up the shapes of tables.

Is dropping 32-26, running db:generate, and then running db:migrate the correct course of action? Generating runs smoothly and gets me back to 26 files, but migrating throws an error `Unknown column 'generated_on' in 'banner_jobs'`. Migration file #25 created a table that definitely had a column `generated_on`.

Dropping 25 as well (where the jobs table was first created) gets a complaint from migrate as well, that `Table 'banner_jobs' already exists` -  I thought it *shouldn't*, because I dropped the migration that created it.

Comment (edenchazard) [reply drizzle/0025_ambiguous_lady_ursula.sql:1]: unfortunately drizzle isn't quite capable of reverting its migrations yet, so you'll have to change them manually

File: server/plugins/queues/shareScrollQueue.ts [hunk 38]
```diff
@@ -1,100 +1,97 @@
...
+        },
+      }
+    );
 
   let shareScrollWorker = createWorker();
 
   if (import.meta.dev) {
-    watch('./workers/shareScrollWorker.js', () => {
-      shareScrollWorker.terminate();
+    watch('/src/workers/shareScrollWorker.js', () => {
```
Comment (edenchazard) [inline server/plugins/queues/shareScrollQueue.ts:49]: was there a reason this was changed?

File: server/plugins/queues/shareScrollQueue.ts [hunk 39]
```diff
@@ -1,100 +1,97 @@
...
-    .on('message', async (message: WorkerResponse) => {
-      if (isJobFinished(message)) {
-        console.log('Job finished with stats: ', message.performanceData);
-
+    .on('failed', async (job: Job) => {
+      // currently, failures not occurring in bannergen, where failures
+      // are always caught and recorded in the perfdata error field,
+      // end up in this block and not recorded in the db.
+      // this includes api failures.
+      // should ALL failures be inserted to db anyway?
```
Comment (edenchazard) [inline server/plugins/queues/shareScrollQueue.ts:62]: ideally yes I think

File: server/plugins/queues/shareScrollQueue.ts [hunk 40]
```diff
@@ -1,100 +1,97 @@
...
+// }
 
-function isJobError(message: WorkerResponse): message is WorkerError {
-  return message.type === 'error';
-}
+// function isJobError(message: WorkerResponse): message is WorkerError {
+//   return message.type === 'error';
+// }
+
+// i ended up not using these but should i?
```
Comment (edenchazard) [inline server/plugins/queues/shareScrollQueue.ts:30]: not unless you have a reason to use them?

File: server/plugins/queues/shareScrollQueue.ts [hunk 41]
```diff
@@ -1,100 +1,97 @@
...
 import { bannerJobsTable } from '~/database/schema';
 import type {
   WorkerInput,
-  WorkerFinished,
-  WorkerResponse,
-  WorkerError,
+  // WorkerFinished,
+  // WorkerResponse,
+  // WorkerError,
+  // since i'm not using these, should they be removed from the types file too?
```
Comment (edenchazard) [inline server/plugins/queues/shareScrollQueue.ts:12]: can you still access the typings in the `on` handlers?

File: server/plugins/queues/shareScrollQueue.ts [hunk 42]
```diff
@@ -1,100 +1,97 @@
...
 } from '~/workers/shareScrollWorkerTypes';
 
 const {
   redis: { host, port },
-  clientSecret,
 } = useRuntimeConfig();
 
 const round = (num: number | null) =>
   num !== null ? parseInt(num.toFixed(0)) : null;
```
Comment (edenchazard) [inline server/plugins/queues/shareScrollQueue.ts:20]: any difference between this and the native `Math.round`?

File: workers/shareScrollWorker.ts [hunk 43]
```diff
@@ -1,35 +1,33 @@
...
 import { ofetch, FetchError } from 'ofetch';
 import type { PerformanceData, WorkerInput } from './shareScrollWorkerTypes';
+import type { Job } from 'bullmq';
 
-const baseBannerWidth = 327;
-const baseBannerHeight = 61;
-const baseCarouselWidth = 106;
+export default async function bannerGen(job: Job) {
+  console.info('Bannergen started for user: ', job.data.user);
+  const perfData: PerformanceData = await generateBannerToTemporary(job.data);
```
Comment (edenchazard) [inline workers/shareScrollWorker.ts:12]: does `job.data` retain the typings?

File: server/plugins/queues/shareScrollQueue.ts (hunk 38 above)
Comment (endulum) [reply server/plugins/queues/shareScrollQueue.ts:35]: Needed an absolute path because a relative path had the code looking for `workers/` somewhere very different.
```
ERROR  Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/src/node_modules/bullmq/dist/cjs/classes/workers/shareScrollWorker.js' imported from /src/node_modules/bullmq/dist/cjs/classes/child-processor.js
hatchery-1  | Did you mean to import "../../../../../workers/shareScrollWorker.js"?
```

Comment (endulum) [reply server/plugins/queues/shareScrollQueue.ts:35]: Oh and if you meant `terminate()` versus `close()`, I'm not sure actually - I was just referencing off the code you provided over Discord and it works. Do we need `terminate()`?
